mantidproject/mantid
# Engg: add Vanadium consideration/corrections

Fixes #12426.

Add vanadium corrections (1: sensitivity, and 2: 'pixel-by-pixel' correction by dividing by a per-bank spline).

**To test:**
- Code review
- See test coverage and that the tests go well on all platforms
- The idea is that validation is as automated as possible with the combination of unit / doc / system tests, but for manual tests see the code examples and comments in the doc tests.

Some explanations on what was done here:

You'll see that the Engg algorithms have now a new input: VanadiumWorkspace, where the vanadium run file is expected. As this modifies the calibration workflow at an early step, all the results change and the tests had to be updated. Besides the new functionality, two additional (intermediate) results were also exposed / added as properties of EnggFocus, EnggCalibrate, EnggCalibrateFull:
- integration results (as a table workspace, one row per spectrum integrated)
- fitted function for every bank (a workspace with the 3 traditional spectra from the algorithm 'Fit' (original data, fitted, difference))

Having the possibility to get and pass these two things from pre-calculated values is useful especially for:
- visualization in the 'diagnostics' gui,
- testing,

It is also convenient from the performance point of view (the vanadium run is a >800MB workspace when loaded, and no need to recalculate for as long as you keep using the same vanadium run).

But this makes me feel that maybe the Vanadium correction step could be turned into an algorithm of its own, a "helper" algorithm like EnggFitPeaks. For example you can see in EnggUtils how I had to pass the parameter "parent" (the parent algorithm to create a child algorithm) over to many functions, which plays the role of the would-be-algorithm 'self'. Most of the new coded added in EnggUtils.py would become the new algorithm, with very little change. I'd discuss this with Saurabh next meeting.
